pkg/frontend/querymiddleware/codec_protobuf.go

```diff
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/prometheus/common/model"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+)
+
+type protobufFormat struct{}
+
+func (f protobufFormat) Name() string {
+	return formatProtobuf
+}
+
+func (f protobufFormat) EncodeResponse(resp *PrometheusResponse) ([]byte, error) {
+	panic("not yet implemented")
+}
+
+func (f protobufFormat) DecodeResponse(buf []byte) (PrometheusResponse, error) {
```

Comment:
[thought] It's quite annoying we have to do all this remapping, which I expect will increase allocations significantly. Was this considered when computing the estimated benefit of protobuf?

Reply:
Yep, all of my benchmarks included the conversion to `PrometheusResponse`.